### DISCUSSION TITLE
How to route /base/[...category]/[slug].js

### DISCUSSION THREAD

**Discussion by justlevine** [vercel/next.js#19854] [Help]
I'm trying to create the following url structure:
`base/[..category].js` => e.g. `/guides/{category}/{subcategory}/{subsubcategory}...` 
`base/category/subcategory/[slug].js` => e.g. `/guides/{category}/{subcategory}/slug`
where I dont know how many categories/subcategories there are.

Tried setting the folder to a catch-all (e.g. `/guides/[...category]/[slug].js), but that's not allowed.

Suggestions?

**Comment by leerob** [comment]
Just do `/guides/[...guide].js`, where guide will now catch category _and_ slug.

https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes

**Reply by justlevine** [reply]
How exactly would this would work, since afaik you cant pass extra params from `get_static_paths()` to `get_static_props()`, and I would need both the `category` and the `slug` to call the correct query and hydrate the component...

**Reply by leerob** [reply]
Check out the docs, there's an example of this 👍 

```
And in the case of /post/a/b, and any other matching path, new parameters will be added to the array, like so:

{ "slug": ["a", "b"] }
```

**Reply by justlevine** [reply]
Yes, the catching part I understand, it's the parsing in `getStaticProps` I'm not getting. `context.params.slug[ slug.length -1 ]` can be either a post or a category, and I need to hydrate the page accordingly (separate gql queries, and I can't combine them since both post() and category() return non-nullable types).

**Reply by leerob** [reply]
Wouldn't the first item be the category, 2nd the subcategory, 3rd is next, and so on? 

`{ "slug": ["cat", "subcat", "thirdcat"] }`

**Reply by justlevine** [reply]
Yes, but I have no way to tell what the last item.is (a cat/subcategory or a post).

Apologies if I'm not being clear. Let me give you some real world examples and what query I need to call in getStaticProps.
/news/ => category query
/news/post-item-1 => post query
/news/local/ => category query
/news/local/post-item-2 => post query
/news/local/press-releases => category query
/news/local/press-releases/post-item-3 => post query.

I can build the paths for these fine in getStaticPaths, but if as you're suggesting I use a single catch-all for both the posts and the categories, then how am I supposed to tell which query to hydrate the page with in getStaticProps?

**Reply by justlevine** [reply]
I _think_ I figured something out, based on your suggestion but I'm not sure if there is a more efficient way to do it. TL;DR in `getStaticProps` i query first for the category, and if that fails I do a second query, passing the successful query type to the component.

```jsx

export default function CategoryAndPost( props ) {
	const { queryType } = props;
	return (
		<Layout>
			<Container>
				{ 'category' === queryType && <CategoryLayout { ...props } /> }
				{ 'post' === queryType && <PostLayout { ...props } /> }
			</Container>
		</Layout>
	);
}

export async function getStaticProps( context ) {
	const slugs = context.params.slug;

	// Query for the category and return.
	const { data } = await client.query( {
		query: EVENTS_BY_CATEGORY_SLUG,
		variables: {
			slug: slugs[ slugs.length - 1 ],		},
	} );
	if ( ! isEmpty( data?.eventsCategory ) ) {
		return {
			props: {
				queryType: 'category',
				posts: data?.eventsCategory?.events?.nodes || [],
			},
			revalidate: 1,
		};
	}
	
	// If not a category, query for post and return.
	const { data: postData } = await client.query( {
		query: EVENT_BY_SLUG_QUERY,
		variables: { slug: slug[ slug.length - 1 ], imageSize: 'FULL' },
	} );
	return {
		props: {
			queryType: 'post',
			post: postData?.event || null,
		},
		revalidate: 1,
	};
}
```

Is there a better way to do this?
